gettsim/taxes/zu_verst_eink/vorsorge.py

```diff
@@ -0,0 +1,218 @@
+
+    By 2020 this rule is by construction the more benefical and the checking against
+    the the pre 2005 rule is not necessary any more. Thus this function is directly
+    assigned to generate the vorsorge data.
+
+    Parameters
+    ----------
+    _altervorsorge_aufwend
+    pflegev_beitr_m
+    ges_krankenv_beitr_m
+    arbeitsl_v_beitr_m
+    eink_st_abzuege_params
+
+    Returns
+    -------
+
+    """
+    out = _altervorsorge_aufwend * 0
+    # 'Basisvorsorge': Health and old-age care contributions are deducted anyway.
+    sonstige_vors = 12 * (
+        pflegev_beitr_m.loc[~kind]
+        + (1 - eink_st_abzuege_params["vorsorge_kranken_minderung"])
+        * ges_krankenv_beitr_m.loc[~kind]
+    )
+    # maybe add unemployment insurance, but do not exceed 1900€.
+    out.loc[~kind] = sonstige_vors.clip(
+        lower=(sonstige_vors + 12 * arbeitsl_v_beitr_m.loc[~kind]).clip(
+            upper=eink_st_abzuege_params["vorsorge_sonstige_aufw_max"]
+        )
+    )
+    out.loc[~kind] += _altervorsorge_aufwend.loc[~kind]
+    return out
+
+
+def _vorsorge_bis_2004(
+    _lohn_vorsorge_bis_2019_single,
+    _lohn_vorsorgeabzug_bis_2019_tu,
+    ges_krankenv_beitr_m,
+    rentenv_beitr_m,
+    ges_krankenv_beitr_m_tu,
+    rentenv_beitr_m_tu,
+    tu_id,
+    gemeinsam_veranlagte_tu,
+    gem_veranlagt,
+    kind,
+    eink_st_abzuege_params,
+):
+
+    out = ges_krankenv_beitr_m * 0
+    out.loc[~gem_veranlagt & ~kind] = _berechne_vorsorge_bis_2004(
+        _lohn_vorsorge_bis_2019_single.loc[~kind],
+        ges_krankenv_beitr_m.loc[~gem_veranlagt & ~kind],
+        rentenv_beitr_m.loc[~gem_veranlagt & ~kind],
+        1,
+        eink_st_abzuege_params,
+    )
+
+    vorsorge_tu = _berechne_vorsorge_bis_2004(
+        _lohn_vorsorgeabzug_bis_2019_tu,
+        ges_krankenv_beitr_m_tu.loc[gemeinsam_veranlagte_tu],
+        rentenv_beitr_m_tu.loc[gemeinsam_veranlagte_tu],
+        2,
+        eink_st_abzuege_params,
+    )
+    out.loc[gem_veranlagt & ~kind] = tu_id[gem_veranlagt].replace(vorsorge_tu)
+    return out
+
+
+def _lohn_vorsorge_bis_2019_single(
+    bruttolohn_m, gemeinsam_veranlagt, eink_st_abzuege_params
+):
+    out = (
+        eink_st_abzuege_params["vorsorge2004_vorwegabzug"]
+        - eink_st_abzuege_params["vorsorge2004_kürzung_vorwegabzug"]
+        * 12
+        * bruttolohn_m.loc[~gemeinsam_veranlagt]
+    ).clip(lower=0)
+    return out
+
+
+def _lohn_vorsorgeabzug_bis_2019_tu(
+    bruttolohn_m_tu, gemeinsam_veranlagte_tu, eink_st_abzuege_params
+):
+    out = 0.5 * (
+        2 * eink_st_abzuege_params["vorsorge2004_vorwegabzug"]
+        - eink_st_abzuege_params["vorsorge2004_kürzung_vorwegabzug"]
+        * 12
+        * bruttolohn_m_tu.loc[gemeinsam_veranlagte_tu]
+    ).clip(lower=0)
+    return out
+
+
+@numpy_vectorize(
+    excluded=["anzahl_erwachsene", "eink_st_abzuege_params"], otypes=[float]
+)
+def _berechne_vorsorge_bis_2004(
```

Comment:
Please break this function apart and remove it. Think about how to write it as operations on the whole series and only make case distinctions where necessary.

Reply:
Works now on the Series level!